### ISSUE TITLE
Solver errors with numeric literals

### ISSUE THREAD

**Issue by jstolarek** [jstolarek/ghc#112]
Overloading causes some issues:

``` haskell
type family Bak a = r | r -> a where
     Bak Int  = Char
     Bak Char = Int
     Bak a    = a

bak :: Bak a -> Bak a
bak x = x

bakapp2 :: Double
bakapp2 = bak 1.0
```

produces:

```
T6018.hs:14:11:
    Couldn't match expected type ‘Double’ with actual type ‘Bak a0’
    The type variable ‘a0’ is ambiguous
    In the expression: bak 1.0
    In an equation for ‘bakapp2’: bakapp2 = bak 1.0
```

Putting a type signature on `1.0` makes the problem go away. Another instance of the same issue:

``` haskell
type family Bar a = r | r -> a where
    Bar Int  = Bool
    Bar Bool = Int
    Bar Bool = Char

bar :: Bar a -> Bar a
bar x = x

barapp2 :: Int
barapp2 = bar 1
```

produces:

```
T6018.hs:125:11:
    Couldn't match expected type ‘Int’ with actual type ‘Bar a0’
    The type variable ‘a0’ is ambiguous
    In the expression: bar 1
    In an equation for ‘barapp2’: barapp2 = bar 1
```

Same thing happens for open type families. Richard says this is a bug in Simon's patch and I won't be able to fix that. Mailed the problem to Simon.

**Comment by jstolarek** [comment]
Blocked by #128

**Comment by jstolarek** [comment]
Occurs check error went away after fixing #128

**Comment by jstolarek** [comment]
Fixes by Simon in 5d8a71bc514f3128d18a2d6e4ec5cf29acb68e6c
